fendermint/vm/interpreter/src/fvm/upgrades.rs

```diff
@@ -1,54 +1,94 @@
 // Copyright 2022-2024 Protocol Labs
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-use std::collections::BTreeMap;
+use std::{collections::BTreeMap, path::Path};
 
-use anyhow::bail;
-use fendermint_vm_core::chainid;
+use anyhow::{bail, Context};
+use fendermint_vm_message::ipc::UpgradeInfo;
 use fvm_ipld_blockstore::Blockstore;
-use fvm_shared::chainid::ChainID;
+use serde::{Deserialize, Serialize};
 use std::collections::btree_map::Entry::{Occupied, Vacant};
 
-use super::state::{snapshot::BlockHeight, FvmExecState};
+use super::state::FvmExecState;
+use fvm_shared::clock::ChainEpoch;
 
-#[derive(PartialEq, Eq, Clone)]
-struct UpgradeKey(ChainID, BlockHeight);
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct UpgradeSchedule {
+    // schedule of upgrades to be executed at given height
+    schedule: BTreeMap<ChainEpoch, UpgradeInfo>,
+}
 
-impl PartialOrd for UpgradeKey {
-    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
-        Some(self.cmp(other))
+impl Default for UpgradeSchedule {
+    fn default() -> Self {
+        Self::new()
     }
 }
 
-impl Ord for UpgradeKey {
-    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
-        if self.0 == other.0 {
-            self.1.cmp(&other.1)
-        } else {
-            let chain_id: u64 = self.0.into();
-            chain_id.cmp(&other.0.into())
+impl UpgradeSchedule {
+    pub fn new() -> Self {
+        Self {
+            schedule: BTreeMap::new(),
+        }
+    }
+
+    pub fn from_file(path: impl AsRef<Path>) -> anyhow::Result<UpgradeSchedule> {
+        let json = std::fs::read_to_string(path).context("failed to read upgrade_info file")?;
+        let schedules = serde_json::from_str::<Vec<UpgradeInfo>>(&json)
+            .context("failed to parse upgrade_info")?;
+
+        let mut us = UpgradeSchedule::new();
+        for upgrade_info in schedules {
+            us.add(upgrade_info)?;
+        }
+
+        Ok(us)
+    }
+
+    pub fn to_file(&self, path: impl AsRef<Path>) -> anyhow::Result<()> {
+        let json = serde_json::to_string_pretty(&self.get_all())
+            .context("failed to serialize upgrade_info")?;
+        std::fs::write(path, json).context("failed to write to upgrade_info file")?;
+
+        Ok(())
+    }
+
+    pub fn add(&mut self, upgrade_info: UpgradeInfo) -> anyhow::Result<()> {
+        match self.schedule.entry(upgrade_info.height) {
+            Vacant(entry) => {
+                entry.insert(upgrade_info);
+                Ok(())
+            }
+            Occupied(_) => {
+                bail!("Upgrade schedule already exists");
+            }
         }
     }
+
+    pub fn get(&self, height: ChainEpoch) -> Option<&UpgradeInfo> {
+        self.schedule.get(&height)
+    }
+
+    pub fn get_all(&self) -> Vec<UpgradeInfo> {
+        self.schedule
+            .values()
+            .cloned()
+            .collect::<Vec<UpgradeInfo>>()
```

Comment:
Maybe return them in reverse, so the top of the file shows useful upcoming info?